### PR TITLE
Parser fixes to support multiple equations/assignments per line.

### PR DESCRIPTION
**Pull request by kblaschke**
Just a few preliminary changes to fix the issue with multiple equations in a single line in shape and wave equations, e.g. "x=1; y=2;".

This PR is only here to make a large number of otherwise broken presets render correctly, as I'm already working on a parser rewrite, as there are so many missing features in it which aren't easy to integrate in the current parser, e.g. having `megabuf(expr) = value;` or `loop(num, expr; expr; ...)`, while the parser currently only supports variables on the left-hand side of a simple equal sign.

### CODE REVIEW

File: src/libprojectM/MilkdropPresetFactory/Parser.cpp [hunk 1]
```diff
@@ -1355,63 +1363,93 @@ int Parser::parse_float(std::istream &  fs, float * float_ptr)
...
+    {
+        return NULL;
+    }
+    if (PARSE_DEBUG)
+    {
+        std::cerr << "parse_per_frame_eqn: parameter \"" << param->name << "\" retrieved (LINE" << line_count << ")"
+                  << std::endl;
+    }
+    /* Make sure parameter is writable */
+    if (param->flags & P_FLAG_READONLY)
```
Comment (revmischa) [inline src/libprojectM/MilkdropPresetFactory/Parser.cpp:1402]: This `if` is a bit different from the one below

File: src/libprojectM/MilkdropPresetFactory/Parser.hpp [hunk 2]
```diff
@@ -29,7 +29,7 @@
 #ifndef _PARSER_H
 #define _PARSER_H
 //#define PARSE_DEBUG 2
-#define PARSE_DEBUG 0
+#define PARSE_DEBUG 2
```
Comment (revmischa) [inline src/libprojectM/MilkdropPresetFactory/Parser.hpp:32]: want to commit this?

**Review by revmischa — APPROVED** [review]
Looks good I guess? hard to tell lol

**Comment by mbellew** [comment]
Consider adding a simple test preset?  These are handy for testing new parser too.

File: src/libprojectM/MilkdropPresetFactory/Parser.cpp (hunk 1 above)
Comment (kblaschke) [reply src/libprojectM/MilkdropPresetFactory/Parser.cpp:1402]: Hum... yeah, I did not try to understand this code at all. This function was just reformatted and I replaced a few `NULL`s with `nullptr`. This line was line 1380 previously, it just moved down a bit.

I think you can see here why it's just better to rewrite this file reader from scratch than trying to "fix" it. Reimplementation will first read all the lines into a map of key/value pairs, then use a key prefix to retrieve init values or code blocks, more or less the same as Milkdrop does it. The existing parser here is just way too complex.

File: src/libprojectM/MilkdropPresetFactory/Parser.hpp (hunk 2 above)
Comment (kblaschke) [reply src/libprojectM/MilkdropPresetFactory/Parser.hpp:32]: Ah, good catch! This should rather be done as a CMake flag, and/or only be enabled in debug builds.

**Comment by kblaschke** [comment]
> Consider adding a simple test preset? These are handy for testing new parser too.

Definitely. [New parser [will be fully unit tested](https://github.com/kblaschke/projectm/blob/9fdebb194692d7c7011f401efcae53bd37d57f72/tests/FileParserTest.cpp), with plenty of presets and equation code to test each feature and function. For the old parser, its simply not worth adding any test cases. For the equation part, I'd personally trust unit tests more than running presets and check the visuals because you don't have any comparison to know how it _should_ look like.

**Comment by mbellew** [comment]
> 

Totally agree.  I would have suggested unit test but i don't know where the unit test app migrated to.

**Comment by kblaschke** [comment]
> Totally agree. I would have suggested unit test but i don't know where the unit test app migrated to.

It's still there, in the top-level "tests" dir, but it'll be replaced with a proper GTest-based framework. Will move all existing "TesstRunner"-based test cases there where it makes sense to keep them.

**Comment by kblaschke** [comment]
I'll just merge this one, as it's a quick, temporary hack to get a few more things working until the new parser is ready.

**Pull request merged**